runtime/lua/vim/lsp/sync.lua

```diff
@@ -0,0 +1,384 @@
+  -- In this case, the first byte change is also at the first byte of firstline
+  if firstline == new_lastline or firstline == lastline then
+    return { line_idx = firstline, byte_idx = 1, char_idx = 1 }
+  end
+
+  local prev_line = prev_lines[firstline]
+  local curr_line = curr_lines[firstline]
+
+  -- Iterate across previous and current line containing first change
+  -- to find the first different byte.
+  -- Note: *about -> a*about will register the second a as the first
+  -- difference, regardless of edit since we do not receive the first
+  -- column of the edit from on_lines.
+  local start_byte_idx = 1
+  for idx = 1, #prev_line + 1 do
+    start_byte_idx = idx
+    if str_byte(prev_line, idx) ~= str_byte(curr_line, idx) then
+      break
+    end
+  end
+
+  -- Convert byte to codepoint if applicable
+  local byte_idx, char_idx = M.align_position(prev_line, start_byte_idx, 'start', offset_encoding)
+
+  -- Return the start difference (shared for new and prev lines)
+  return { line_idx = firstline, byte_idx = byte_idx, char_idx = char_idx }
+end
+
+---@private
+--- Finds the last line and byte index of the differences between prev and current buffer.
+--- Normalized to the next codepoint.
+--- prev_end_range is the text range sent to the server representing the changed region.
+--- curr_end_range is the text that should be collected and sent to the server.
+--
+---@param prev_lines table list of lines
+---@param curr_lines table list of lines
+---@param start_range table
+---@param lastline integer
+---@param new_lastline integer
+---@param offset_encoding string
+---@returns (int, int) end_line_idx and end_col_idx of range
+function M.compute_end_range(prev_lines, curr_lines, start_range, firstline, lastline, new_lastline, offset_encoding)
+  -- If firstline == new_lastline, the first change occured on a line that was deleted.
+  -- In this case, the last_byte...
+  if firstline == new_lastline then
+      return { line_idx = (lastline - new_lastline + firstline), byte_idx = 1, char_idx = 1 }, { line_idx = firstline, byte_idx = 1, char_idx = 1 }
+  end
+  if firstline == lastline then
+      return { line_idx = firstline, byte_idx = 1, char_idx = 1 }, { line_idx = new_lastline - lastline + firstline, byte_idx = 1, char_idx = 1 }
+  end
+  -- Compare on last line, at minimum will be the start range
+  local start_line_idx = start_range.line_idx
+
+  -- lastline and new_lastline were last lines that were *not* replaced, compare previous lines
+  local prev_line_idx = lastline - 1
+  local curr_line_idx = new_lastline - 1
+
+  local prev_line = prev_lines[lastline - 1]
+  local curr_line = curr_lines[new_lastline - 1]
+
+  local prev_line_length = #prev_line
+  local curr_line_length = #curr_line
+
+  local byte_offset = 0
+  -- Buffer has increased in line count
+  -- if new_lastline > lastline then
+  -- -- Buffer has decreased in line count
+  -- elseif new_lastline < lastline then
+  -- elseif
+  -- end
```

Comment:
This looks a bit as if it is a leftover from iterating on the code. I guess it can be removed?